query/expression.go

```diff
@@ -144,19 +144,66 @@ func (expr *metricFetchExpression) Evaluate(context EvaluationContext) (value, e
 }
 
 func (expr *functionExpression) Evaluate(context EvaluationContext) (value, error) {
-	switch expr.functionName {
+
+	name := expr.functionName
+	switch name {
 	case "+":
-		return evaluateBinaryOperation(context, expr.functionName, expr.arguments,
-			func(left, right float64) float64 { return left + right })
+		fallthrough
 	case "-":
-		return evaluateBinaryOperation(context, expr.functionName, expr.arguments,
-			func(left, right float64) float64 { return left - right })
-	case "/":
-		return evaluateBinaryOperation(context, expr.functionName, expr.arguments,
-			func(left, right float64) float64 { return left / right })
+		fallthrough
 	case "*":
-		return evaluateBinaryOperation(context, expr.functionName, expr.arguments,
-			func(left, right float64) float64 { return left * right })
+		fallthrough
+	case "/":
+		arguments := make([]value, len(expr.arguments))
+		var err error
+		for i := range arguments {
+			arguments[i], err = expr.arguments[i].Evaluate(context)
+			if err != nil {
+				return nil, err
+			}
+		}
+		// Evaluation of a binary operator:
+		if len(arguments) != 2 {
+			return nil, errors.New(fmt.Sprintf("Function `%s` expects 2 operands but received %d (%+v)", name, len(arguments), arguments))
+		}
+		left := arguments[0]
+		right := arguments[1]
+		operatorMap := map[string]func(float64, float64) float64{
+			"+": func(x, y float64) float64 { return x + y },
+			"-": func(x, y float64) float64 { return x - y },
+			"*": func(x, y float64) float64 { return x * y },
+			"/": func(x, y float64) float64 { return x / y },
+		}
+		return evaluateBinaryOperation(context, name, left, right, operatorMap[name])
+	case "aggregate.sum":
+		fallthrough
+	case "aggregate.mean":
+		fallthrough
+	case "aggregate.min":
+		fallthrough
+	case "aggregate.max":
+		arguments := make([]value, len(expr.arguments))
+		var err error
+		for i := range arguments {
+			arguments[i], err = expr.arguments[i].Evaluate(context)
+			if err != nil {
+				return nil, err
+			}
+		}
+		funMap := map[string]aggregate{
+			"aggregate.sum":  sumAggregate,
+			"aggregate.mean": meanAggregate,
+			"aggregate.min":  minAggregate,
+			"aggregate.max":  maxAggregate,
+		}
+		if len(arguments) != 1 {
+			return nil, errors.New(fmt.Sprintf("Function `%s` expected 2 operands but received %d (%+v)", name, len(arguments), arguments))
+		}
+		list, err := arguments[0].toSeriesList(context.Timerange)
+		if err != nil {
+			return nil, err
+		}
+		return seriesListValue(aggregateBy(list, aggregateMap[funMap[name]], expr.groupBy)), nil
```

Comment:
Yes, have the enum => aggregate mapping occur in `aggregateBy`.

I still think we should export the enums and `aggregateBy`, not for package-external visibility, but for package-internal identification of things that are meant to be used. Use of unexported things should be limited to the files in which they are defined. Though, if we start doing this it _may_ be useful to start splitting these into packages.